lib/plugins/cors.js

```diff
@@ -73,66 +120,234 @@ function matchOrigin(req, origins) {
+        }
+
+        // restify router has a sig of function found(err, route, context) {
+        // but we don't use route or context in this case. for simplicity
+        // sake, the user facing preflightStrategy should just return an
+        // array of methods as the second param.
+        findMethodsFn(req, res, function found(err, resMethods) {
+
+            if (err) {
+                if (err instanceof errs.ResourceNotFoundError ||
+                    err.restCode === 'ResourceNotFound') {
+                    return next(new errs.BadRequestError(errMsg));
+                }
+            }
+
+            // look for it first on res, then on the returned value, then
+            // finally fall back to empty array
+            var supportedMethods = res.methods || resMethods || [];
+
+            // ensure the requested method is within this list of supported
+            // methods.
+            if (supportedMethods.indexOf(reqMethod) === -1) {
+                return next(new errs.MethodNotAllowedError(errMsg2));
             }
 
-            res.setHeader(AC_EXPOSE_HEADERS, headers.join(', '));
+            // the request looks good. set headers and we're good to go.
+            setCommonHeaders(res, reqOrigin, matchedOrigin);
+            res.setHeader(AC_ALLOW_HEADERS, allowHeaders.join(', '));
+            res.setHeader(AC_ALLOW_METHODS, supportedMethods);
+            res.setHeader(AC_MAX_AGE, preflightMaxAge);
+
+            // end the request, return false so restify will stop processing
+            // the request.
+            res.writeHead(200);
+            res.end();
+            return next(false);
+        });
+    }
+
+
+
+    // sets headers for use in CORS requests
+    function headers(req, res, next) {
+
+        // 1) if no origin header incoming request, move on, set no headers.
+        var reqOrigin = req.headers.origin;
+
+        if (!reqOrigin) {
+            return next();
         }
 
-        res.once('header', corsOnHeader);
-        next();
+        // 2) if no matched origin found, move on, set no headers.
+        var matchedOrigin = matchOrigin(reqOrigin, origins);
+
+        if (!matchedOrigin) {
+            return next();
+        }
+
+        // if match was found, let's set some headers.
+        setCommonHeaders(res, reqOrigin, matchedOrigin);
+        res.setHeader(AC_EXPOSE_HEADERS, exposeHeaders.join(', '));
+
+        return next();
     }
 
-    return (restifyCORSSimple);
+
+    // set common cors headers
+    function setCommonHeaders(res, reqOrigin, matchedOrigin) {
+        if (credentials === true) {
+            // when credentials are used, not allowed to use *!
+            res.setHeader(AC_ALLOW_ORIGIN, reqOrigin);
+            res.setHeader(AC_ALLOW_CREDS, 'true');
+            res.setHeader(STR_VARY, STR_ORIGIN);
+        } else {
+            // for non credentialed requests
+            if (matchedOrigin === '*') {
+                res.setHeader(AC_ALLOW_ORIGIN, '*');
```

Comment:
I think there's some semantics around * and the vary header. I can't recall exactly, but I suspect this is the reason for all the special casing of * in this PR.